### PR TITLE
Improve networking extended test reliability

### PR DESCRIPTION
**Pull request by marun**
See the individual commit messages for details.

### CODE REVIEW

**Comment by marun** [comment]
[testonlyextended][extended:networking]

**Comment by marun** [comment]
This PR will includes #8099 until it merges to ensure that test runs don't fail due to ineffective node readiness detection.

**Comment by marun** [comment]
@stevekuznetsov A review would be welcome.  I'm not happy with how I'm having to disable error checking for statements that I need more nuanced control over, and if you have suggestions how to accomplish the same thing more elegantly I'm all ears.

File: hack/dind-cluster.sh [hunk 1]
```diff
@@ -324,11 +324,12 @@ os::provision::install-cmds ${DEPLOYED_ROOT}"
 }
 
 function nodes-are-ready() {
-    local node_count=$(${DOCKER_CMD} exec -t "${MASTER_NAME}" bash -c "\
+  # Skip the SDN node since nothing is intended to be scheduled on it.
+  local node_count=$(${DOCKER_CMD} exec -t "${MASTER_NAME}" bash -c "\
 KUBECONFIG=${DEPLOYED_CONFIG_ROOT}/openshift.local.config/master/admin.kubeconfig \
```
Comment (stevekuznetsov) [inline hack/dind-cluster.sh]: The formatting seems to be all wrong here -- makes it a little hard to parse.

File: hack/dind-cluster.sh [hunk 2]
```diff
@@ -324,11 +324,12 @@ os::provision::install-cmds ${DEPLOYED_ROOT}"
...
-oc get nodes | grep Ready | wc -l")
-    node_count=$(echo "${node_count}" | tr -d '\r')
-    test "${node_count}" -ge "${NODE_COUNT}"
+oc get nodes | grep -v SchedulingDisabled | grep -v NotReady | grep Ready | wc -l")
```
Comment (stevekuznetsov) [inline hack/dind-cluster.sh]: I have a couple of comments about this invocation:
 - why are we `docker exec`-ing into the container? Can't the local OpenShift client ping the server in the container over the network?
 - we should pass in the configuration using `oc get nodes --config=...` instead of using the `$KUBECONFIG` envar
 - Instead of the complicated multi-level `grep` that's going on here, a `jsonpath` or `go-template` structure would be better to grab the correct field, followed by a `wc -l` for counting.

File: hack/dind-cluster.sh [hunk 3]
```diff
@@ -324,11 +324,12 @@ os::provision::install-cmds ${DEPLOYED_ROOT}"
...
+  node_count=$(echo "${node_count}" | tr -d '\r')
```
Comment (stevekuznetsov) [inline hack/dind-cluster.sh]: Is this test running on Windows? How are we getting carriage returns?

File: test/extended/networking.sh [hunk 4]
```diff
@@ -7,9 +7,19 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-if [[ "${OPENSHIFT_QUIET_OUTPUT:-false}" != "true" ]]; then
+CONFIG_ROOT="${OPENSHIFT_CONFIG_ROOT:-}"
+
+# CI mode (deploying clusters for each plugin) should be noisy by
+# default to aid in debugging.
+if [[ "${CONFIG_ROOT}" = "" ]]; then
```
Comment (stevekuznetsov) [inline test/extended/networking.sh]: The idiomatic way to check if a string is empty is:
```sh
if [[ -z "${var:-}" ]]; then
    # var is empty
fi
```

File: test/extended/networking.sh [hunk 5]
```diff
@@ -7,9 +7,19 @@ set -o errexit
...
+  DEFAULT_QUIET=false
+else
+  DEFAULT_QUIET=true
+fi
+
+if [[ "${OPENSHIFT_QUIET_OUTPUT:-${DEFAULT_QUIET}}" != "true" ]]; then
```
Comment (stevekuznetsov) [inline test/extended/networking.sh]: All you really care about here is that `$DEFAULT_QUIET` is not (equal to the string) `"true"`.

The logic here is doing two default steps, so you could really change this `if` statement to:
```sh
if [[ -n "${OPENSHIFT_QUIET_OUTPUT:-${CONFIG_ROOT:-}}" ]]; then
    # do work
fi
```

File: test/extended/networking.sh [hunk 6]
```diff
@@ -85,50 +99,81 @@ function save-artifacts() {
...
-# Any non-zero exit code from any test run invoked by this script
-# should increment TEST_FAILURE so the total count of failed test runs
-# can be returned as the exit code.
-TEST_FAILURES=0
-function test-osdn-plugin() {
+function deploy-cluster() {
   local name=$1
   local plugin=$2
   local isolation=$3
+  local log_dir=$4
```
Comment (stevekuznetsov) [inline test/extended/networking.sh]: Unclear if it's needed, but some sort of `if [[ $# -lt 4 ]]; then os::log::error "deploy-cluster expected four arguments, got $#"; fi` statement might be useful as a guard here as we have `set -o nounset` on.

File: test/extended/networking.sh [hunk 7]
```diff
@@ -85,50 +99,81 @@ function save-artifacts() {
...
+  # Disable error checking to have a chance to save logs and artifacts
+  # if cluster deployment fails.
+  set +e
+
+  # Restart instead of start to ensure that an existing cluster is
+  # always torn down.
+  ${CLUSTER_CMD} restart
+  exit_status=$?
+
+  if [ "${exit_status}" = "0" ]; then
```
Comment (stevekuznetsov) [inline test/extended/networking.sh]: The idiomatic way to do guarded commands when you intend to execute different code based on the output is to simply wrap the command into an `if` statement itself:
```sh
if ! ${CLUSTER_CMD} restart; then
    # failed
fi
```

Comment (stevekuznetsov) [inline test/extended/networking.sh]: The space after `!` token is necessary here and the code will not execute correctly if it is missing -- something I've found to be a subtle nuisance.

File: test/extended/networking.sh [hunk 8]
```diff
@@ -256,13 +300,12 @@ else
   # Docker-in-docker is not compatible with selinux
   disable-selinux
 
-  os::log::info "Ensuring that previous test cluster is shut down"
-  ${CLUSTER_CMD} stop
-
-  test-osdn-plugin "subnet" "redhat/openshift-ovs-subnet" "false"
+  # Ignore deployment errors for a given plugin to allow other plugins
+  # to be tested.
+  test-osdn-plugin "subnet" "redhat/openshift-ovs-subnet" "false" || true
```
Comment (stevekuznetsov) [inline test/extended/networking.sh]: Whenever a `|| true` shows up, the question must be asked -- why is failure here something we should tolerate without any sort of action acknowledging it?

**Comment by stevekuznetsov** [comment]
@marun I've left some comments -- hope they are helpful. As you seem to be doing work on Bash scripts pretty often, I'd like to suggest that you work [ShellCheck](http://www.shellcheck.net/) linting into your workflow. It will not be added to the CI flow as it's not currently available in the EPEL/RHEL repositories, but the tool I have found to be very useful for pointing out potential pitfalls, especially nuanced ones that are easy to overlook.

File: test/extended/networking.sh (hunk 5 above)
Comment (stevekuznetsov) [inline test/extended/networking.sh]: Or, a little less convoluted:
```sh
if [[ -n "${OPENSHIFT_QUIET_OUTPUT:-}" || -n "${CONFIG_ROOT:-}" ]]; then
    # do work
fi
```

File: test/extended/networking.sh (hunk 6 above)
Comment (marun) [inline test/extended/networking.sh]: I don't think it's required in this case - test-managed deployment with this method will always want to retain logs.

File: test/extended/networking.sh (hunk 5 above)
Comment (marun) [inline test/extended/networking.sh]: How does what you've specified allow a developer to suppress verbose output if OPENSHIFT_CONFIG_ROOT is null?

File: test/extended/networking.sh (hunk 7 above)
Comment (marun) [inline test/extended/networking.sh]: That's very helpful, thank you.  I got stuck on wanting to return the status code from the failing command but really all I need to do is branch on failure.

File: hack/dind-cluster.sh (hunk 2 above)
Comment (marun) [inline hack/dind-cluster.sh]: All fair points, but the change in question has already merged in a different PR (it was causing job failures).  I'll follow up in a separate PR.

File: hack/dind-cluster.sh (hunk 3 above)
Comment (marun) [inline hack/dind-cluster.sh]: The output from the docker invocation includes the carriage returns.

**Comment by marun** [comment]
@stevekuznetsov I wonder if the increasing complexity of origin's bash scripting isn't a sign that we should consider switching to a dynamic language like python or ruby.  afaik bash is impervious to test-driven development and I think that can only put an upper bound on the complexity of problems we can solve in ci.

**Comment by marun** [comment]
cc: @openshift/networking

**Comment by danwinship** [comment]
One problem I'd noticed with OPENSHIFT_QUIET_OUTPUT is that I always had to do "echo $?" afterward to see if all the tests passed or if there was some failure several screens back. (With non-quiet this isn't a problem because you see the "`+ exit 0`" there.) So it would be nice to explicitly print the final success/failure result before exiting

**Comment by stevekuznetsov** [comment]
> So it would be nice to explicitly print the final success/failure result before exiting

If my changes for our traps ever go in, you'll always get this for free.

**Comment by danwinship** [comment]
the rest lgtm

File: test/extended/networking.sh (hunk 5 above)
Comment (stevekuznetsov) [inline test/extended/networking.sh]: The code above allows a developer to pass some non-null string in using *either* `$OPENSHIFT_QUIET_OUTPUT` or `$OPENSHIFT_CONFIG_ROOT` -- I believe this is an equivalent logical statement to what you had written, except for the fact that the *content* of `$OPENSHIFT_QUIET_OUTPUT` is no longer important, simply the presence of some non-null string is `true` -- that's idiomatic for Bash.

For instance:
```sh
function existing_logic() {
	local CONFIG_ROOT="${OPENSHIFT_CONFIG_ROOT:-}"
	local DEFAULT_QUIET

	if [[ -z "${CONFIG_ROOT}" ]]; then
	    DEFAULT_QUIET=false
	else
	    DEFAULT_QUIET=true
	fi

    echo -n "existing logic: "
	if [[ "${OPENSHIFT_QUIET_OUTPUT:-${DEFAULT_QUIET}}" != "true" ]]; then
	    echo "true"
	else
	    echo "false"
	fi
}

function proposed_logic() {
    echo -n "proposed logic: "
    if [[ -n "${OPENSHIFT_QUIET_OUTPUT:-}" || ! -n "${OPENSHIFT_CONFIG_ROOT:-}" ]]; then
        echo "true"
    else
        echo "false"
    fi
}

echo "both unset"
existing_logic
proposed_logic

echo "config root unset, quiet output true"
OPENSHIFT_QUIET_OUTPUT=true existing_logic
OPENSHIFT_QUIET_OUTPUT=true proposed_logic

echo "config root unset, quiet output false"
OPENSHIFT_QUIET_OUTPUT=false existing_logic
OPENSHIFT_QUIET_OUTPUT=false proposed_logic

echo "config root set, quiet output unset"
OPENSHIFT_CONFIG_ROOT='/some/path' existing_logic
OPENSHIFT_CONFIG_ROOT='/some/path' proposed_logic

echo "config root set, quiet output true"
OPENSHIFT_CONFIG_ROOT='/some/path' OPENSHIFT_QUIET_OUTPUT=true existing_logic
OPENSHIFT_CONFIG_ROOT='/some/path' OPENSHIFT_QUIET_OUTPUT=true proposed_logic

echo "config root set, quiet output false"
OPENSHIFT_CONFIG_ROOT='/some/path' OPENSHIFT_QUIET_OUTPUT=false existing_logic
OPENSHIFT_CONFIG_ROOT='/some/path' OPENSHIFT_QUIET_OUTPUT=false proposed_logic
```

The output is:
```
both unset
existing logic: true
proposed logic: true
config root unset, quiet output true
existing logic: false
proposed logic: true
config root unset, quiet output false
existing logic: true
proposed logic: true
config root set, quiet output unset
existing logic: false
proposed logic: false
config root set, quiet output true
existing logic: false
proposed logic: true
config root set, quiet output false
existing logic: true
proposed logic: true
```

**Comment by stevekuznetsov** [comment]
>  I wonder if the increasing complexity of origin's bash scripting isn't a sign that we should consider switching to a dynamic language like python or ruby

I think this very much might be the case, but I don't really know enough to have an authoritative opinion on it. I think Bash makes a lot of sense for something like `test-cmd.sh`, but that the tooling we write for our Bash could definitely be in a different (better-suited) language, wrapped up by a Bash library so the two can interface. I believe @Miciah is familiar with how this looked in v2.

File: hack/dind-cluster.sh [hunk 9]
```diff
@@ -327,7 +327,7 @@ function nodes-are-ready() {
   # Skip the SDN node since nothing is intended to be scheduled on it.
   local node_count=$(${DOCKER_CMD} exec -t "${MASTER_NAME}" bash -c "\
 KUBECONFIG=${DEPLOYED_CONFIG_ROOT}/openshift.local.config/master/admin.kubeconfig \
-oc get nodes | grep -v ${SDN_NODE_NAME} | grep -v NotReady | grep Ready | wc -l")
+oc get nodes | grep -v SchedulingDisabled | grep -v NotReady | grep Ready | wc -l")
```
Comment (stevekuznetsov) [inline hack/dind-cluster.sh]: I'd still strongly recommend `jsonpath` or `go-templates` for  extracting the values that you want to count -- this approach is fragile as the output is not stable.

**Comment by marun** [comment]
re-[testonlyextended]

File: hack/dind-cluster.sh (hunk 9 above)
Comment (danwinship) [inline hack/dind-cluster.sh]: Yeah, ok. So this will print a list of all ready, schedulable nodes:

    oc get nodes --template '{{range $item := .items}}{{if not .spec.unschedulable}}{{range .status.conditions}}{{if eq .type "Ready"}}{{if eq .status "True"}}{{printf "%s\n" $item.metadata.name}}{{end}}{{end}}{{end}}{{end}}{{end}}'

**Comment by marun** [comment]
re-[testonlyextended]

File: hack/dind-cluster.sh (hunk 9 above)
Comment (marun) [inline hack/dind-cluster.sh]: Kudos to figuring it out, but ugh, go-templates are ugly as sin.  :/

Comment (danwinship) [inline hack/dind-cluster.sh]: eh, it makes sense after a while :)
(and some of the ugliness is just because the data model is complicated; eg, we have to find the "Ready" condition in the arbitrarily-ordered array of conditions and then check its value)

Comment (stevekuznetsov) [inline hack/dind-cluster.sh]: I'd suggest `jsonpath` over `go-template`. It usually is much cleaner.

Comment (stevekuznetsov) [inline hack/dind-cluster.sh]: However, it's cleaner because the logic exists outside of the object introspection.

File: hack/dind-cluster.sh [hunk 10]
```diff
@@ -104,9 +104,15 @@ function check-selinux() {
...
+DIND_IMAGE="openshift/dind"
 IMAGE_REGISTRY="${OPENSHIFT_TEST_IMAGE_REGISTRY:-}"
+if [[ -n "${IMAGE_REGISTRY}" ]]; then
+  DIND_IMAGE="${IMAGE_REGISTRY}/${DIND_IMAGE}"
+fi
 IMAGE_TAG="${OPENSHIFT_TEST_IMAGE_TAG:-}"
-DIND_IMAGE="${IMAGE_REGISTRY}openshift/dind${IMAGE_TAG}"
+if [[ -n "${IMAGE_TAG}" ]]; then
+  DIND_IMAGE="${DIND_IMAGE}:${IMAGE_TAG}"
+fi
```
Comment (pravisankar) [inline hack/dind-cluster.sh]: This script got a lot of config options, either documenting under 'OpenShift Configuration' section or moving config params to hack/dind-config.sh will be helpful.

Comment (marun) [inline hack/dind-cluster.sh]: It's hard to argue against better docs, but the image caching configuration is removed by #8520 and most of the remaining options are already shared with vagrant (contrib/vagrant/provision-config.sh).  Is there something in particular that prompted this comment?

File: test/extended/networking.sh (hunk 5 above)
Comment (marun) [inline test/extended/networking.sh]: I appreciate your attempts to educate me, but I find 'idiomatic' bash to be less readable.  Most programming I've done uses == / != so having seemingly arbitrary flags (-z, -n) in their place doesn't make sense to me.  Is there a functional reason for this kind of divergence?

Comment (marun) [inline test/extended/networking.sh]: My other challenge is internalizing that idiomatic bash would appear to require flags intended to be boolean as binary (not present or present) rather than depending on value.  I don't find it intuitive that OPENSHIFT_QUIET_OUTPUT=false be equivalent to OPENSHIFT_QUIET_OUTPUT=true.  That might make sense in bash, but in most non-bash applications a parameter value means something.

Comment (marun) [inline test/extended/networking.sh]: I think there's a logical error in the proposed logic.  Why should the the following test indicate that verbose output was called for? 

    -n ${OPENSHIFT_QUIET_OUTPUT:-}

Here's the truth table of the proposed logic:
 
|   config root  | quiet           | verbose  |
| ------------- |:-------------:| -----:|
| unset      |  unset | true  |
| unset      | set      |  true  |
| set      | unset      | false |
| set      | set      |   true |

Even were that logic fixed, the truth table for a binary quiet would be:

|   config root      | quiet           | verbose  |
| ------------- |:-------------:| -----:|
| unset      |  unset | true  |
| unset      | set      |  false  |
| set      | unset      | false |
| set      | set      |   false |

Assuming I want to be able to enable logging when config root is set, I think quiet would need to be a ternary value:

|   config root      | quiet           | verbose  |
| ------------- |:-------------:| -----:|
| unset      |  unset | true  |
| unset      | true      |   false |
| unset   | false      |   true |
| set     | unset      |   false |
| set     | true      |   false |
| set      | false     |   true|

Clearly this is too complicated, though.  I'm going to simplify so that verbose output is controlled by a single binary flag.

**Comment by marun** [comment]
I've submitted a cleanup of nodes-are-ready in #8522

File: test/extended/networking.sh (hunk 5 above)
Comment (stevekuznetsov) [inline test/extended/networking.sh]: > Why should the the following test indicate that verbose output was called for? 

Simply because in Bash, the presence of an environment variable is equivalent to a boolean `true` -- if I set `$OPENSHIFT_QUIET_OUTPUT`, I want the output to be quiet.

I understand your gripe about `-n` and `-z`, but they become second-nature fairly quickly. I think the best reason to favor idiomatic Bash is that it's idiomatic. When I write a script that requires a environment variable to equal the string `"true"`, someone else using it will see a boolean environment variable and assume that it's used idiomatically. I recently ran into this when I added jUnit functionality into `hack/test-go.sh`. I personally always pass in `$VAR=true` just to be extra clear (but never use `$VAR=false`), but Clayton used `$VAR=1` when he consumed my script, as that's idiomatic. The script was robust to both approaches as it used the idiomatic Bash expression.

While I agree that Bash is a wily beast, I think our *best* weapon against it is using an idiomatic and consistent approach to writing it. 

If you do `$ man true` or `$ man false` you will also see that `true` and `false` are *not* built-ins, but *programs*. This means that if someone is attempting to do something similar to what you were doing but accidentally evaluated the variable or evaluated `true` when doing the assignment, they'd see incorrect behavior. 

Furthermore, `[`, `[[`, and `if` have an inverse evaluation of numbers-as-booleans, that is `0` is true and `1` (technically, any non-zero number) is false. This is idiomatic for checking the return codes of commands, but not for storing booleans as integers like in C. Even furthermore, the `((` construct evaluates booleans as integers correctly, that is `0` is false and `1` is true. All this being said, trying to represent booleans in Bash using non-Bash-isms is usually harder to read and reason about (which is why I made the comments originally, I had to write a truth table in order to actually understand what your code was doing). 

When I was writing the `os::cmd` suite, I thought a lot about boolean handling in Bash, and through much trial and error I settled on the idiomatic manner. I'm not sure that we should deviate from it, as idiomatic Bash is our best bet at writing scripts many people can consume without having to think hard about the API.

File: test/extended/networking.sh [hunk 11]
```diff
@@ -189,7 +214,7 @@ ${OPENSHIFT_INSTANCE_PREFIX:-openshift}"
     fi
     ;;
   *)
-    if [ "${CONFIG_ROOT}" != "" ]; then
+    if [ -n "${CONFIG_ROOT}" ]; then
```
Comment (stevekuznetsov) [inline test/extended/networking.sh]: We should prefer `[[` over `[`

Comment (stevekuznetsov) [inline test/extended/networking.sh]: Also, to guard against `$CONFIG_ROOT` being unset, we can default it: `if [[ -n "${CONFIG_ROOT:-}" ]]; then ...`

**Comment by eparis** [comment]
Anything that uses grep against the text output of oc commands is a bad idea... I realize that we try not to break people scripts, but it has happened and will happen again. grep is brittle.

**Comment by marun** [comment]
@eparis I've tended to automate based on how I use the cli, and I think go templates are overkill for that, but I see how it's the way to go for automation.

File: test/extended/networking.sh (hunk 11 above)
Comment (marun) [inline test/extended/networking.sh]: re: [[ vs [ - Done.

Regarding CONFIG_ROOT, it's an alias to OPENSHIFT_CONFIG_ROOT and it's initialized on line 202: 

https://github.com/openshift/origin/pull/8506/files#diff-71ec4266e3a35d109f56f03dc3cd3ed9R202

**Comment by marun** [comment]
@stevekuznetsov Anything else?  I'd really like to get this merged so I can have enough detail to diagnose flakes I'm seeing in ci.

cc: @eparis Please merge when @stevekuznetsov is happy.

**Comment by stevekuznetsov** [comment]
LGTM

File: test/extended/networking.sh (hunk 5 above)
Comment (marun) [inline test/extended/networking.sh]: On Thu, Apr 14, 2016 at 8:38 PM, Steve Kuznetsov <notifications@github.com>
wrote:
>
> Simply because in Bash, the presence of an environment variable is
> equivalent to a boolean true -- if I set $OPENSHIFT_QUIET_OUTPUT, I want
> the output to be quiet.
>

My confusion wasn't around quiet being set = true, unset = false.  I didn't
understand why the logic you proposed was using `-n $OPENSHIFT_QUIET_OUTPUT`
as one indication that verbose output should be configured.  Did you mean
to use `-z`?

**Comment by marun** [comment]
@stevekuznetsov Thank you for your diligence in reviewing and educating me, much appreciated!

**Comment by eparis** [comment]
approved and [merge]

**Comment by marun** [comment]
re-[test]
